exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/SqlConverter.java

```diff
@@ -0,0 +1,364 @@
+          .addContext("SQL Query", formatSQLParsingError(sql, e.getPos()))
+          .build(logger);
+    }
+
+  }
+
+  public SqlNode validate(final SqlNode parsedNode) {
+    try {
+      SqlNode validatedNode = validator.validate(parsedNode);
+      return validatedNode;
+    } catch (RuntimeException e) {
+      throw UserException
+          .validationError(e)
+          .message(isInnerQuery ?
+              "Failure validating a view your query is dependent upon." : "Failure validating query.")
+          .addContext("SQL query", sql)
+          .build(logger);
+    }
+  }
+
+  public RelDataType getOutputType(SqlNode validatedNode) {
+    return validator.getValidatedNodeType(validatedNode);
+  }
+
+  public JavaTypeFactory getTypeFactory() {
+    return typeFactory;
+  }
+
+  public SqlOperatorTable getOpTab() {
+    return opTab;
+  }
+
+  public RelOptCostFactory getCostFactory() {
+    return costFactory;
+  }
+
+  public SchemaPlus getRootSchema() {
+    return rootSchema;
+  }
+
+  public SchemaPlus getDefaultSchema() {
+    return defaultSchema;
+  }
+
+  private class DrillValidator extends SqlValidatorImpl {
+    protected DrillValidator(SqlOperatorTable opTab, SqlValidatorCatalogReader catalogReader,
+        RelDataTypeFactory typeFactory, SqlConformance conformance) {
+      super(opTab, catalogReader, typeFactory, conformance);
+    }
+  }
+
+  private static class DrillTypeSystem extends RelDataTypeSystemImpl {
+
+    @Override
+    public int getDefaultPrecision(SqlTypeName typeName) {
+      switch (typeName) {
+      case CHAR:
+      case BINARY:
+      case VARCHAR:
+      case VARBINARY:
+        return 65536;
+      default:
+        return super.getDefaultPrecision(typeName);
+      }
+    }
+
+    @Override
+    public int getMaxNumericScale() {
+      return 38;
+    }
+
+    @Override
+    public int getMaxNumericPrecision() {
+      return 38;
+    }
+
+  }
+
+  public RelNode toRel(
+      final SqlNode validatedNode) {
+    final RexBuilder rexBuilder = new RexBuilder(typeFactory);
+    if (planner == null) {
+      planner = new VolcanoPlanner(costFactory, settings);
+      planner.setExecutor(new DrillConstExecutor(functions, util, settings));
+      planner.clearRelTraitDefs();
+      planner.addRelTraitDef(ConventionTraitDef.INSTANCE);
+      planner.addRelTraitDef(DrillDistributionTraitDef.INSTANCE);
+      planner.addRelTraitDef(RelCollationTraitDef.INSTANCE);
+    }
+
+    final RelOptCluster cluster = RelOptCluster.create(planner, rexBuilder);
+    final SqlToRelConverter sqlToRelConverter =
+        new SqlToRelConverter(new Expander(), validator, catalog, cluster, DrillConvertletTable.INSTANCE);
+
+    sqlToRelConverter.setTrimUnusedFields(false);
+    sqlToRelConverter.enableTableAccessConversion(false);
+    final RelNode rel = sqlToRelConverter.convertQuery(validatedNode, false, true);
```

Comment:
sqlToRelConverter.convertQuery(): the last param is boolean top, indicating whether the query is top-level. In case that it's called from Expander.expandView(), should we set this boolean flag to be false?

In the original code path, where PlannerImpl.ViewExpanderImpl is used, it seems to set the boolean flag as false.